### DISCUSSION TITLE
More info when catching exceptions

### DISCUSSION THREAD

**Discussion by thebyohazard** [nanodbc/nanodbc#362] [Q&A]
Is there any way to get more information about what is going wrong when calling my stored procedure?
I am getting the error "Indicator variable required but not supplied" and I am unsure as to why.
I am calling a stored procedure in a MSSQL Server 2008 Database from a program on an ubuntu 20.04 computer using nanodbc.
Here is an example of my code:

```
try {
	nanodbc::connection connection(get_connection_string(),sqlTimeout);
	nanodbc::statement statement(connection);
	prepare(statement,"{CALL " + storedProcedureName_ + "(?,?,?,?,?,?)};");
	statement.timeout(sqlTimeout);
	int nextArg = 0;
	statement.bind(nextArg++,param1_.c_str());
	statement.bind(nextArg++,param2_.c_str());
	char param3[50];
	statement.bind_strings(nextArg++,param3,(size_t)50,(size_t)1,nanodbc::statement::PARAM_OUT);
	char param4[50];
	statement.bind_strings(nextArg++,param4,(size_t)50,(size_t)1, nanodbc::statement::PARAM_OUT);
	char param5[50];
	statement.bind_strings(nextArg++,param5,(size_t)50,(size_t)1, nanodbc::statement::PARAM_OUT);
	char param6[50];
	statement.bind_strings(nextArg++,param6,(size_t)50,(size_t)1, nanodbc::statement::PARAM_OUT);
	statement.execute();
	
	classVar_ = param3;
	classVar2_ = param4;
	classVar3_ = param5;
	classVar4_ = param6;
}
catch(const std::exception& e) {
	cout << "Could not call procedure: " << e.what();
}
```

When the stored procedure fails, it prints out:

> Could not call procedure: /path/to/nanodbc.cpp:1684: 2200: [Microsoft][ODBC Driver 18 for SQL Server]Indicator variable required but not supplied 

From what I can tell, if my procedure returns null for any of the output parameters, then this error will be displayed.
However, at the end of the stored procedure I do an isnull around each of the variables to ensure that all output params are not null.
Can this message be displayed if the procedure times out? Is there any other reason for this message to be displayed? 
Can I see what the procedure does return even if it does not return all the variables? Thanks!

**Comment by mloskot** [comment]
https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlfetch-function says:

```
| 22002 
| Indicator variable required but not supplied 
| NULL data was fetched into a column whose StrLen_or_IndPtr set by SQLBindCol
(or SQL_DESC_INDICATOR_PTR set by SQLSetDescField or SQLSetDescRec)
was a null pointer.
```

which suggests to me indicator buffer is not being set correctly and is null pointer. You will need to debug and investigate the what binding does for your case.

I don't know any other reasons for this message.

Try stepping through `recent_error` as suggested here https://github.com/nanodbc/nanodbc/discussions/310#discussioncomment-2611008  to verify if there is more diagnostics available. The (complex) errors handling needs to be improved indeed https://github.com/nanodbc/nanodbc/issues/361#issuecomment-1316683428

**Reply by thebyohazard** [reply]
Thanks for your help!

**Comment by thebyohazard** [comment]
So in my case, my stored procedure had a try/catch block that would rollback the transaction if any errors were raised. The output parameters were not being set in the catch block, so if an exception was raised in the stored proc before the variables were set, then at the end of the call, the parameters would still be null, and thus the indicator variable required warning would be displayed. 

It is worth noting that the indicator variable message would be the message of the c++ exception, rather than the message of the raised error in the stored procedure.

So for future people who find themselves here, make sure all code paths of your stored procedure set values for output parameters. Even if there is no exception raised and you don't use the parameters in c++, it will still issue this warning if you do not set your parameters.
